Introduce TI EHRPWM/EPWM driver

natto1784 opened the pull request:
Dependency: https://github.com/zephyrproject-rtos/zephyr/pull/106755/changes

This patchset adds
- TI EHRPWM driver
- Nodes for AM243x EVM and AM62L EVM
- Changed generic syscon's offset (reg) type from uint16_t to uint32_t to allow configuration for larger offsets

natto1784 commented:
Hello, it has been a while since I added this. I would like some opinions on the clock_control driver of this patchset. As you can see we are handling multiple compatible strings in the same driver. This requires creation of multiple bindings. Is there a better way to do this>

natto1784 commented:
CC: @gramsay0 @Ayush1325 @glneo - Please feel free to review if and when you get some time.

glneo commented on drivers/clock_control/clock_control_ti_syscon.c:
```diff
@@ -0,0 +1,132 @@
+
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_AM64_COMPAT)
+static const struct ti_syscon_gate_clk_id_data am64_clk_ids[] = {
+	{0, 0}, {0, 1}, {0, 2}, {0, 3}, {0, 4}, {0, 5}, {0, 6}, {0, 7}, {0, 8},
```
Any good way to encode this data into DT so that we don't need to keep device specific data here in the driver?

Ayush1325 reviewed (commented):
Can you add the device to the dts of whatever board you are testing with?

natto1784 commented:
not stale but waiting, due to PR dependency

natto1784 committed:
dts: ti,control-module: add epwm-tbclk space-specifier

Add a specifier space for configuring timebase clock for TI's Enhanced PWM
via syscon.

Signed-off-by: Amneesh Singh <amneesh@ti.com>

natto1784 committed:
dts: ti: am64x: add TISCI clock controller

Add the TISCI clock controller with the node label k3_clks.

Signed-off-by: Amneesh Singh <amneesh@ti.com>

natto1784 committed:
am243x_evm: am2434: r5f0_0: add support for TI EPWM

Add DT nodes for TI's Enhanced PWM. This requires addition of a syscon node
as well to configure the timebase clock.

Signed-off-by: Amneesh Singh <amneesh@ti.com>

natto1784 committed:
am62l_evm: am62l3: a53: add support for TI EPWM

Add DT nodes for TI's Enhanced PWM. This requires addition of a syscon node
as well to configure the timebase clock.

Signed-off-by: Amneesh Singh <amneesh@ti.com>

natto1784 committed:
drivers: pwm: introduce ti_am3352_ehrpwm

This is the driver for EPWM/EHRPWM module of TI SoCs. It only uses Action
Qualifier submodule currently.

Signed-off-by: Amneesh Singh <amneesh@ti.com>

natto1784 commented on drivers/syscon/syscon.c:
```diff
@@ -33,7 +33,7 @@ static int syscon_generic_get_base(const struct device *dev, uintptr_t *addr)
 	return 0;
 }
 
-static int syscon_generic_read_reg(const struct device *dev, uint16_t reg, uint32_t *val)
```
moved to another PR (https://github.com/zephyrproject-rtos/zephyr/pull/106755), and added that as a dependency to this.

natto1784 commented on dts/vendor/ti/am64x_main.dtsi:
```diff
@@ -523,4 +538,94 @@
+
+	main_epwm0: pwm@23000000 {
+		compatible = "ti,am3352-ehrpwm";
+		reg = <0x23000000 0x100>;
+		tbclk = <&main_mmr_cfg0 0x4130 0>;
```
Copying my reply from the other thread:

These regions in K3 device have registers that are used to gate clocks and power and such. More than one peripheral may need to access registers in the same register space. I figured that instead of mapping space for one integer per driver, it may be better to just use one syscon node to define the entire region and then that space can be used by any driver like in the aforementioned patch.

natto1784 committed:
drivers: pwm: ti_am3352_ehrpwm: add PWM event callback support

Add interrupt-driven PWM event callback functionality to the TI AM3352
EHRPWM driver. Note that only one event can be enabled at a time, hence we
require all callbacks to have same event trigger or none at all.

Signed-off-by: Amneesh Singh <amneesh@ti.com>

natto1784 commented:
I never realized that ISR callback support for PWM non-capture drivers was added after I wrote this driver, so adding support for that in a new patch, will also solve merge conflicts soon.

glneo commented on drivers/pwm/pwm_ti_am3352_ehrpwm.c:
```diff
@@ -21,16 +27,21 @@ LOG_MODULE_REGISTER(ti_ehrpwm);
-	volatile uint16_t AQCTLB; /**< AQ Control Register for Output B, offset: 0x18 */
-	volatile uint16_t AQSFRC; /**< AQ Software Force Register, offset: 0x1A */
-	volatile uint16_t AQCSFRC /**< AQ Software Continuous Force Register, offset: 0x1C */;
+	volatile uint16_t TBCTL;   /**< Time-Base Control Register, offset: 0x00 */
```
Next time don't worry about keeping the comment starts all aligned, it just adds churn as seen here when you add a longer register. Can't easily see what is added/removed when the whole block is changed.

natto1784 replied:
that makes sense, thanks.

natto1784 commented:
Fixed conflicts, reviews got dismissed: @glneo. Thanks.
